weed/s3api/s3acl/acl_helper_test.go

```diff
@@ -0,0 +1,220 @@
+package s3acl
+
+import (
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/seaweedfs/seaweedfs/weed/s3api/s3_constants"
+	"github.com/seaweedfs/seaweedfs/weed/s3api/s3account"
+	"github.com/seaweedfs/seaweedfs/weed/s3api/s3err"
+	"testing"
+)
+
+func TestParseAclHeaders(t *testing.T) {
+	accountManager := &s3account.AccountManager{
+		IdNameMapping: map[string]string{
+			s3account.AccountAdmin.Id:     s3account.AccountAdmin.Name,
+			s3account.AccountAnonymous.Id: s3account.AccountAnonymous.Name,
+		},
+		EmailIdMapping: map[string]string{
+			s3account.AccountAdmin.EmailAddress:     s3account.AccountAdmin.Id,
+			s3account.AccountAnonymous.EmailAddress: s3account.AccountAnonymous.Id,
+		},
+	}
+
+	//good value
+	grants := make([]*s3.Grant, 0)
+	validHeaderValue := `uri="http://acs.amazonaws.com/groups/global/AllUsers", id="anonymous", emailAddress="admin@example.com"`
+	errCode := ParseCustomAclHeader(validHeaderValue, s3_constants.PermissionFullControl, &grants)
+	if errCode != s3err.ErrNone {
+		t.Fatal(errCode)
+	}
+	_, errCode = ValidateAndTransferGrants(accountManager, grants)
+	if errCode != s3err.ErrNone {
+		t.Fatal(errCode)
+	}
+
+	//bad case: acl header format error
+	grants = make([]*s3.Grant, 0)
+	formatErrCase := `uri, id="anonymous", emailAddress="admin@example.com"`
+	errCode = ParseCustomAclHeader(formatErrCase, s3_constants.PermissionFullControl, &grants)
+	if errCode != s3err.ErrInvalidRequest {
+		t.Fatal(errCode)
+	}
+
+	//bad case: email not exists
+	grants = make([]*s3.Grant, 0)
+	badCaseOfEmail := `uri="http://acs.amazonaws.com/groups/global/AllUsers", id="anonymous", emailAddress="admin@example1.com"`
+	errCode = ParseCustomAclHeader(badCaseOfEmail, s3_constants.PermissionFullControl, &grants)
+	if errCode != s3err.ErrNone {
+		t.Fatal(errCode)
+	}
+	_, errCode = ValidateAndTransferGrants(accountManager, grants)
+	if errCode != s3err.ErrInvalidRequest {
+		t.Fatal(errCode)
+	}
+
+	//bad case: account id not exists
+	grants = make([]*s3.Grant, 0)
+	badCaseOfAccountId := "uri=\"http://acs.amazonaws.com/groups/global/AllUsers\", id=\"xxxxxx\", emailAddress=\"admin@example.com\""
+	errCode = ParseCustomAclHeader(badCaseOfAccountId, s3_constants.PermissionFullControl, &grants)
+	if errCode != s3err.ErrNone {
+		t.Fatal(errCode)
+	}
+	_, errCode = ValidateAndTransferGrants(accountManager, grants)
+	if errCode != s3err.ErrInvalidRequest {
+		t.Fatal(errCode)
+	}
+
+	//bad case: group url not valid
+	grants = make([]*s3.Grant, 0)
+	badCaseOfURL := "uri=\"http://acs.amazonaws.com/groups/global/AllUsers111xxxx\", id=\"anonymous\", emailAddress=\"admin@example.com\""
+	errCode = ParseCustomAclHeader(badCaseOfURL, s3_constants.PermissionFullControl, &grants)
+	if errCode != s3err.ErrNone {
+		t.Fatal(errCode)
+	}
+	_, errCode = ValidateAndTransferGrants(accountManager, grants)
+	if errCode != s3err.ErrInvalidRequest {
+		t.Fatal(errCode)
+	}
+}
+
+func TestGrantEquals(t *testing.T) {
+	testCases := map[bool]bool{
```

Comment:
there are only 2 entries in this map. I do not think this test works as expected.

Reply:
ok, I'll add more tests

Reply:
@chrislusf done